modules/codegen/src/main/scala/com/twilio/guardrail/generators/Http4sServerGenerator.scala

```diff
@@ -0,0 +1,561 @@
+                optional(scalaParam)(tpe)
+              case param"$_: Option[$tpe] = $_" =>
+                optional(scalaParam)(tpe)
+              case param"$_: Iterable[$tpe]" =>
+                multi(scalaParam)(tpe)
+              case param"$_: Iterable[$tpe] = $_" =>
+                multi(scalaParam)(tpe)
+              case _ => required(scalaParam)(argType)
+            }
+        }
+      } yield directives
+
+    def bodyToHttp4s(operationId: String, body: Option[ScalaParameter]): Target[Option[Enumerator.Generator]] =
+      Target.pure(
+        body.map {
+          case ScalaParameter(_, _, paramName, _, argType) =>
+            enumerator"${Pat.Var(paramName)} <- req.as[${argType}](E, ${Term.Name(s"${operationId}Decoder")})"
+        }
+      )
+
+    def headersToHttp4s: List[ScalaParameter] => Target[List[Enumerator]] =
+      directivesFromParams(
+        arg => {
+          case t"String" => Target.pure(enumerator"${Pat.Var(arg.paramName)} <- E.pure(req.headers.get(${arg.argName.toLit}.ci).map(_.value).get)")
+          case tpe =>
+            Target.pure(
+              enumerator"${Pat.Var(arg.paramName)} <- E.pure(req.headers.get(${arg.argName.toLit}.ci).map(_.value).flatMap(Json.fromString(_).as[$tpe].toOption).get)"
+            )
+        },
+        arg => _ => Target.error(s"Unsupported Iterable[${arg}"),
+        arg => _ => Target.error(s"Unsupported Option[Iterable[${arg}]]"),
+        arg => {
+          case t"String" => Target.pure(enumerator"${Pat.Var(arg.paramName)} <- E.pure(req.headers.get(${arg.argName.toLit}.ci).map(_.value))")
+          case tpe =>
+            Target.pure(
+              enumerator"${Pat.Var(arg.paramName)} <- E.pure(req.headers.get(${arg.argName.toLit}.ci).map(_.value).flatMap(Json.fromString(_).as[$tpe].toOption))"
+            )
+        }
+      )
+
+    def qsToHttp4s(operationId: String): List[ScalaParameter] => Target[Option[Pat]] =
+      params =>
+        directivesFromParams(
+          arg => _ => Target.pure(p"${Term.Name(s"${operationId.capitalize}${arg.paramName.value.capitalize}Matcher")}(${Pat.Var(arg.paramName)})"),
+          arg => _ => Target.pure(p"${Term.Name(s"${operationId.capitalize}${arg.paramName.value.capitalize}Matcher")}(${Pat.Var(arg.paramName)} @ _*)"),
+          arg => _ => Target.pure(p"${Term.Name(s"${operationId.capitalize}${arg.paramName.value.capitalize}Matcher")}(${Pat.Var(arg.paramName)})"),
+          arg => _ => Target.pure(p"${Term.Name(s"${operationId.capitalize}${arg.paramName.value.capitalize}Matcher")}(${Pat.Var(arg.paramName)})")
+        )(params).map {
+          case Nil => Option.empty
+          case x :: xs =>
+            Some(xs.foldLeft[Pat](x) { case (a, n) => p"${a} +& ${n}" })
+      }
+
+    def formToHttp4s: List[ScalaParameter] => Target[List[Enumerator]] =
+      directivesFromParams(
+        arg => _ => Target.pure(enumerator"${Pat.Var(arg.paramName)} = urlForm.values.get(${arg.argName.toLit}).get.head"),
```

Comment:
If a parameter doesn't match in akka-http, the routing layer falls down to the next method. If no routes match, a 404 is generated. It seems as though using `.get` and `head` would throw exceptions, which would cause a server error instead of user error.

Without pushing form handling into the routing layer, I don't immediately see a better solution than what you have, though possibly introducing a mandatory `400 Bad Request` in a method's response and using that in the `fold` after doing all the argument parsing safely would be an acceptable strategy.

Reply:
Additionally, to work around not having literal `enum` support (and to support more complex domain types), it's currently acceptable to use `x-scala-type` like this: https://github.com/twilio/guardrail/blob/master/modules/sample/src/main/resources/formData.yaml#L34

I noticed that we didn't have any generator/compiler tests for that file, so I added the following to `build.sbt`:
```diff
diff --git a/build.sbt b/build.sbt
index 2cc58dd..2cd17a8 100644
--- a/build.sbt
+++ b/build.sbt
@@ -1,3 +1,5 @@
+import scala.sys.process._
+
 val projectName = "guardrail-root"
 name := projectName
 organization in ThisBuild := "com.twilio"
@@ -61,12 +63,16 @@ fullRunTask(
   --client --specPath modules/sample/src/main/resources/custom-header-type.yaml --outputPath modules/sample/src/main/scala --packageName tests.customTypes.customHeader.client.akkaHttp --framework akka-http
   --server --specPath modules/sample/src/main/resources/custom-header-type.yaml --outputPath modules/sample/src/main/scala --packageName tests.customTypes.customHeader.server.http4s --framework http4s
   --server --specPath modules/sample/src/main/resources/custom-header-type.yaml --outputPath modules/sample/src/main/scala --packageName tests.customTypes.customHeader.server.akkaHttp --framework akka-http
+  --client --specPath modules/sample/src/main/resources/formData.yaml --outputPath modules/sample/src/main/scala --packageName form.client.akkaHttp --tracing --framework akka-http
+  --client --specPath modules/sample/src/main/resources/formData.yaml --outputPath modules/sample/src/main/scala --packageName form.client.http4s --tracing --framework http4s
+  --server --specPath modules/sample/src/main/resources/formData.yaml --outputPath modules/sample/src/main/scala --packageName form.server.akkaHttp --tracing --framework akka-http
+  --server --specPath modules/sample/src/main/resources/formData.yaml --outputPath modules/sample/src/main/scala --packageName form.server.http4s --tracing --framework http4s
 """.replaceAllLiterally("\n", " ").split(' ').filter(_.nonEmpty): _*
 )
 
 artifact in (Compile, assembly) := {
   val art = (artifact in (Compile, assembly)).value
-  art.copy(`classifier` = Some("assembly"))
+  art.withClassifier(Some("assembly"))
 }
 
 addArtifact(artifact in (Compile, assembly), assembly)
diff --git a/project/build.properties b/project/build.properties
index 133a8f1..0cd8b07 100644
--- a/project/build.properties
+++ b/project/build.properties
@@ -1 +1 @@
-sbt.version=0.13.17
+sbt.version=1.2.3
```
frustratingly, I hit some sort of bug in sbt's macro compilation for `runExample`, but upgrading to the latest sbt seems to have fixed it. (though all the same plugin versions were crosspublished to 1.x, so that was fortunate)

Reply:
Introduced a `400 Bad Request` if a parameter doesn't match. Added support for more complex domain types. Added tests for `formData.yaml` file.